### ISSUE TITLE
Some alerts would be misreporting when prometheus has mutl shards

### ISSUE THREAD

**Issue by yccharles** [prometheus-operator/kube-prometheus#1744]
**What happened?**
I apply a Prometheus using shards=3 like this 
```
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
metadata:
    prometheus: k8s
  name: k8s
  namespace: monitoring
spec:
  shards: 3
   ...
```
Than i got some alert from those expressions, such as
```
    - alert: KubeSchedulerDown
      expr: |
        absent(up{job="kube-scheduler"} == 1)
```
```
    - alert: KubeAPIDown
      expr: |
        absent(up{job="apiserver"} == 1)
```

This is caused by Prometheus sharding, because some of Prometheus instances do not monitor these components. But those alert expressions was running in each  Prometheus instances . 

Here is an example: 
We has 3 instance apiserver running in cluster. 

- apiserver1: 10.xx.xx.158
- apiserver2: 10.xx.xx.174
- apiserver3:  10.xx.xx.65

We also have 3 shards of Prometheus instances. 

- prometheus-k8s-0                                         3/3     Running   0          3d1h
- prometheus-k8s-shard-1-0                                 3/3     Running   0          31h
- prometheus-k8s-shard-2-0                                 3/3     Running   0          31h

And apiserver1  and apiserver2 would be monitoring by prometheus-k8s-0  , apiserver3  would be monitoring by prometheus-k8s-shard-1-0 . 

Although 3 instances apiservice is still alive, But An alarm is initiat by prometheus-k8s-shard-2-0 .

**Did you expect to see some different?**

May be those expressions should be monitoring by thanos-rules.

**Environment**
kube-prometheus release-0.9

**Comment by hartfordfive** [comment]
Hi all.  Any news on this?  Thank you.

**Comment by simonpasquier** [comment]
> May be those expressions should be monitoring by thanos-rules.

This is correct. If you configure sharding with the Prometheus operator, the expectation is that you leverage Thanos query + ruler to offload the rule evaluations from the Prometheus instances to Thanos Ruler. Because sharding is based on the target labels, the operator can't predict which Prometheus shard is going to scrape which metric.

```
Prometheus shard #0 <-----------+
                                 \
Prometheus shard #1 <-------------+- Thanos query <-- Thanos Ruler
...                              /
Prometheus shard #N <-----------+
```

**Comment by funkypenguin** [comment]
@simonpasquier , could you expand a little on how the above can be achieved? Would we need to disable the built-in PrometheusRules (_like the one which provides `KubeSchedulerDown`_), or is simply enabling Ruler and telling it to use existing PrometheusRules (_which seems to be the default helm chart configuration_) going to be enough?

Thank you!

**Comment by simonpasquier** [comment]
@funkypenguin the second option: deploying the Prometheus rules on Thanos Ruler instead of Prometheus.

**Comment by alelomonaco** [comment]
@simonpasquier I have the same issue, I enabled thanos query and ruler, I specified to use existingSDConfigmap and the same replicaLabel for query and rule, but KubeApiDown alert still fire. Prometheus operator is configured with 2 replicas and 2 shard. Looking at single Prometheus instance up{job="apiserver"}, metric is detected by shards 0 and not detected by shards 1. The same shard balancing method is used for every metric, but those one use "absent" formula get into this error. Could you please  describe more in depth which configurations solve this issue?

Thank you very much!

**Comment by Aslan-Liu** [comment]
> @funkypenguin the second option: deploying the Prometheus rules on Thanos Ruler instead of Prometheus.

I have the same question. If I just enable Thanos Ruler and use rules from Prometheus, each prometheus instance still evaluate each rule and fire alerts in it. Isn't it? 

How could we do?

**Comment by omriarieli** [comment]
> > @funkypenguin the second option: deploying the Prometheus rules on Thanos Ruler instead of Prometheus.
> 
> I have the same question. If I just enable Thanos Ruler and use rules from Prometheus, each prometheus instance still evaluate each rule and fire alerts in it. Isn't it?
> 
> How could we do?

i did this by telling the prometheus shards to not look at any rules like so
```yaml
  prometheusSpec:
    ruleSelector:
      matchExpressions:
        - key: nothing
          operator: Exists
```
the tradeoff for this though is if you are using grafana you will not see the alert under alerting in Mimir / Cortex / Loki

**Comment by leehosu** [comment]
I stumbled on that issue today. Any news on it?

**Comment by leehosu** [comment]
The solution to this issue is to have Thanos Ruler monitor the alert rules instead of Prometheus. You can configure this using the `prometheus.prometheusSpec.ruleSelector` feature.
